### PR TITLE
DAOS-14494 test: Add a a new data cache mode for dfuse.

### PR DESCRIPTION
**Pull request by ashleypittman**
Support a new "open to close" cache mode for dfuse data.  This mode
allows caching whilst the file is open but does not allow cache re-use
after close.  This is similar to disabling caching in terms of impact but
does allow MAP_SHARED to work on dfuse mounts without needing to
set a data cache timeout.

Add a new test for this mode.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by jolivier23 — DISMISSED** [review]

File: docs/user/filesystem.md [hunk 1]
```diff
@@ -666,19 +667,21 @@ to be set to 0 or off, except dentry-dir-time which defaults to dentry-time
 | dfuse-dentry-time       | How long directory entries are cached                                  |
 | dfuse-dentry-dir-time   | How long dentries are cached, if the entry is itself a directory       |
 | dfuse-ndentry-time      | How long negative dentries are cached                                  |
-| dfuse-data-cache        | Data caching enabled, duration or ("on"/"true"/"off"/"false")          |
+| dfuse-data-cache        | Data caching enabled, duration or ("on"/"true"/"off"/"false"/"otoc")   |
```
Comment (mchaarawi) [inline docs/user/filesystem.md:670]: otoc sound weird and not intuitive on what it is and why is it needed. i don't have a better suggestion though. maybe "on_mmap" or something like that? not requesting a change just an observation.

File: src/client/dfuse/ops/open.c [hunk 2]
```diff
@@ -72,6 +72,13 @@ dfuse_cb_open(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi)
 		} else {
 			prefetch = true;
 		}
+	} else if (ie->ie_dfs->dfc_data_otoc) {
+		/* Open to close caching, this allows the use of shared mmap */
+		fi_out.direct_io  = 0;
+		fi_out.keep_cache = 0;
+
+		if (fi->flags & O_DIRECT)
+			fi_out.direct_io = 1;
```
Comment (mchaarawi) [inline src/client/dfuse/ops/open.c:81]: if-else statement looks better here for direct_io than assigning then overwriting in case of O_DIRECT

**Review by mchaarawi — APPROVED** [review]

File: docs/user/filesystem.md (hunk 1 above)
Comment (ashleypittman) [reply docs/user/filesystem.md:670]: "close to open" consistency is a well known term so I thought "open to close" caching made sense in that regard - and it's clear then what effect this setting will have.

File: src/tests/ftest/util/dfuse_utils.py [hunk 3]
```diff
@@ -423,7 +423,7 @@ def start_dfuse(test, dfuse, pool=None, container=None, **params):
     if pool:
         params['pool'] = pool.identifier
     if container:
-        params['cont'] = container.uuid
+        params['cont'] = container.identifier
```
Comment (daltonbohning) [inline src/tests/ftest/util/dfuse_utils.py:426]: Thanks. FWIW, I still have work in progress to update 99% of ftest to use labels

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/util/dfuse_utils.py (hunk 3 above)
Comment (daltonbohning) [reply src/tests/ftest/util/dfuse_utils.py:426]: Luckily I have a PR to make this same change and there wasn't fallout.

**Pull request merged**